python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,211 @@
+    """An output file implementation for the FsspecFileIO
+
+    Args:
+        location(str): A URI to a file location
+        fs: An fsspec filesystem instance
+    """
+
+    def __init__(self, location: str, fs):
+        self._fs = fs
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._fs.info(self.location)
+        if size := object_info.get("Size"):
+            return size
+        elif size := object_info.get("size"):
+            return size
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._fs.lexists(self.location)
+
+    def create(self, overwrite: bool = False):
+        """Create an output stream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object
+
+        Raises:
+            FileExistsError: If the file already exists at the location and overwrite is set to False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Returns a new FsspecInputFile for the location at `self.location`"""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, properties: Properties):
+        self._scheme_to_fs = {}
+        self._scheme_to_fs.update(SCHEME_TO_FS)
+        self._fs_cache: Dict[str, AbstractFileSystem] = {}
+        super().__init__(properties=properties)
+
+    def new_input(self, location: str) -> FsspecInputFile:
+        """Get an FsspecInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecInputFile: An FsspecInputFile instance for the given location
+        """
+        uri = urlparse(location)
+
+        if uri.scheme not in self._scheme_to_fs:
+            raise ValueError(f"No registered filesystem for scheme: {uri.scheme}")
+
+        if cached_fs := self._fs_cache.get(uri.scheme):
```

Comment:
I think it would be better to cache this using `functools`: https://docs.python.org/3/library/functools.html#functools.cache Moving to a separate function `def _get_fs(scheme: str)` would just cache for each fo the schemas.

Reply:
The code of initializing an FS is also defined both in the InputFile, OutputFile and delete, would be nice if we can consolidate this as well.

Reply:
Bugbear flagged this with:

> B019: Use of functools.lru_cache or functools.cache on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.

The solution I used was to just initialize a cache in the class init and add the instance method to it. This localizes the cache so that it's destroyed with the instance.
```py
self.get_fs: Callable = lru_cache(self._get_fs)
```

Let me know what you think.